### ISSUE TITLE
[DTensor] High overhead in optimizer step when training with DTensor

### ISSUE THREAD

**Issue by KingsleyLiu-NV** [pytorch/pytorch#109101]
### 🐛 Describe the bug

Hi all, we are implementing tensor parallel convolution based on DTensor, and faces an performance issue that may be related DTensor. In our use case, we replicate model weights across GPUs using DTensor, and pass parameters of distributed model to native PyTorch optimizer. This leads to really high overhead in the optimizer step:

```python
# Correct, Poor Performance
model = ...
model = distribute_module(model, mesh, _data_parallel_fn, input_fn=None, output_fn=None)
optimizer = torch.optim.Adam(model.parameters(), lr=1e-4, amsgrad=False)
```

If we pass parameters of model to optimizer before calling `distribute_module` , the training process is wrong in that no model weights are actually updated:
```python
# Wrong, Good Performance
model = ...
optimizer = torch.optim.Adam(model.parameters(), lr=1e-4, amsgrad=False)
model = distribute_module(model, mesh, _data_parallel_fn, input_fn=None, output_fn=None)
```
I am not sure if native optimizer can handle DTensor performantly and what is the recommended practice to train with DTensor.

Below is the simple script to reproduce the aforementioned performance issue. The order of line50 (`distribute_module`) and line51 (`torch.optim.Adam`) can be reversed to cover what I mentioned above.

```python
import torch
import torch.nn as nn
​
import torch.distributed as dist
from torch.distributed._tensor import (
  DTensor,
  DeviceMesh,
  distribute_tensor,
  distribute_module,
  Shard,
  Replicate
)
import torch.multiprocessing as mp
import os
import time
​
WORLD_SIZE = 4
ITER_TIME = 20
​
def _data_parallel_fn(
    name: str,
    module: nn.Module,
    device_mesh: DeviceMesh,
) -> None:
    for name, param in module.named_parameters():
        dist_spec = ([Replicate()])
        dist_param = torch.nn.Parameter(
            distribute_tensor(param, device_mesh, dist_spec)
        )
        name = '_'.join(name.split('.'))
        module.register_parameter(name, dist_param)
​
def demo_data_parallel(rank, world_size):
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12355"
​
    dist.init_process_group("nccl", rank=rank, world_size=world_size)
    
    in_shape = [20, 32]
    output_shape = [20, 5]
​
    device = torch.device("cuda", rank)
    torch.cuda.set_device(device)
    mesh = DeviceMesh("cuda", torch.arange(world_size))
​
    model = torch.nn.Linear(32, 5)
    nn.init.ones_(model.weight)
    nn.init.zeros_(model.bias)
​
    model = distribute_module(model, mesh, _data_parallel_fn, input_fn=None, output_fn=None)
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-4, amsgrad=False)
​
    x = torch.randn(*in_shape).to(device).requires_grad_()
    y_grad = torch.randn(*output_shape).to(device)
    x = distribute_tensor(x, mesh, [Shard(0)])
    y_grad = distribute_tensor(y_grad, mesh, [Shard(0)])
​
    start = time.time()
    for i in range(ITER_TIME):
        print(f"---------------{i}--------------------")
        torch.cuda.nvtx.range_push("model_iter"+str(i))
​
        torch.cuda.nvtx.range_push("forward")
        y = model(x)
        torch.cuda.synchronize(device)
        torch.cuda.nvtx.range_pop()
​
        torch.cuda.nvtx.range_push("zero_grad")
        optimizer.zero_grad()
        torch.cuda.synchronize(device)
        torch.cuda.nvtx.range_pop()
​
        torch.cuda.nvtx.range_push("backward")
        y.backward(y_grad)
        torch.cuda.synchronize(device)
        torch.cuda.nvtx.range_pop()
​
        torch.cuda.nvtx.range_push("optimizer_step")
        optimizer.step()
        torch.cuda.synchronize(device)
        torch.cuda.nvtx.range_pop()
​
        torch.cuda.synchronize(device)
        torch.cuda.nvtx.range_pop()
    end = time.time()
    max_reserved_memory = torch.cuda.max_memory_reserved(device)
    max_allocated_memory = torch.cuda.max_memory_allocated(device)
​
    print(f"rank {rank}, {ITER_TIME} iterations, latency {(end - start)/ITER_TIME*1000} ms, max reserved {max_reserved_memory/1024/1024/1024:8.2f} GiB, max allocated {max_allocated_memory/1024/1024/1024:8.2f} GiB")
    dist.destroy_process_group()
​
if __name__ == "__main__":
    mp.spawn(demo_data_parallel, args=(WORLD_SIZE,), nprocs=WORLD_SIZE, join=True)
```

### Versions

docker image: [nvcr.io/nvidia/pytorch:23.08-py3](http://nvcr.io/nvidia/pytorch:23.08-py3)

torch version: '2.1.0a0+29c30b1'

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by KingsleyLiu-NV** [comment]
@wanchaol @fduwjj FYI

**Comment by awgu** [comment]
@KingsleyLiu-NV I am just passing by to mention: I think you might have forgotten to switch the order in the "Wrong, Good Performance" example. IIUC, should it be:
```
# Wrong, Good Performance
model = ...
optimizer = torch.optim.Adam(model.parameters(), lr=1e-4, amsgrad=False)
model = distribute_module(model, mesh, _data_parallel_fn, input_fn=None, output_fn=None)
```

**Comment by KingsleyLiu-NV** [comment]
> @KingsleyLiu-NV I am just passing by to mention: I think you might have forgotten to switch the order in the "Wrong, Good Performance" example. IIUC, should it be:
> 
> ```
> # Wrong, Good Performance
> model = ...
> optimizer = torch.optim.Adam(model.parameters(), lr=1e-4, amsgrad=False)
> model = distribute_module(model, mesh, _data_parallel_fn, input_fn=None, output_fn=None)
> ```

Thanks for pointing out. I have edited the description accordingly.

**Comment by wanchaol** [comment]
@KingsleyLiu-NV yep DTensor currently has some CPU overhead, we are improving the performance via: directly reducing its CPU overhead (some of them already landed in latest main) 2. enable torch.compile that can fully get rid of its overhead. The main use case today with DTensor in eager is Tensor Parallel, where it only partition certain linear layers. 

Before these overhead reduction come into place, It's recommend to only apply DTensor to part of your model instead of the whole model (i.e. if you only want to partition convolution you can just shard that conv layer and leave the rest untouched by only apply `distribute_module` to conv layer), this would help reduce the overhead.

**Comment by KingsleyLiu-NV** [comment]
> Before these overhead reduction come into place, It's recommend to only apply DTensor to part of your model instead of the whole model (i.e. if you only want to partition convolution you can just shard that conv layer and leave the rest untouched by only apply distribute_module to conv layer), this would help reduce the overhead.

This cannot solve the performance issue. I benchmark DTensor performance on a single-convolution-layer model and the overhead of optimizer step still dominates significantly, while forward and backward with DTensor is of acceptable overhead.

I think the script I provide in this issue can demonstrate perfectly my point: it is a model with very few layers, and optimizer step with DTensor is disproportionately slow.

**Comment by wanchaol** [comment]
> I think the script I provide in this issue can demonstrate perfectly my point: it is a model with very few layers, and optimizer step with DTensor is disproportionately slow.

Hmm if it's optimizer only it's out of my expectation, I'll take a look into this soon.

**Comment by fduwjj** [comment]
@KingsleyLiu-NV Do you happen to have the trace? If not, we can try to collect on our end.

**Comment by wanchaol** [comment]
@KingsleyLiu-NV debugged a bit on this and found that there's some caching bug/regression in the dtensor caching layer, which causes the optimizer cache misses and thus the high overhead you observed. I submitted a fix on this already and will try to land this soon

**Comment by wanchaol** [comment]
@KingsleyLiu-NV Note that after this fix, optimizer step would greatly reduce its overhead, but you might notice in the trace that it's still not small enough, this is because you are doing data parallel training, for data parallel training DTensor would need to do gradient averaging similar to DDP, but the gradient averaging (partial -> replicate) is lazily executed until the optimizer step, this is by the design of DTensor. One can eagerly do the allreduce by installing gradient hooks to DTensor parameter so that the partial gradient convert to replicate gradient immediately after the backward gradient is available, or even do some fusion on it, this totally depend on how you want to approach it.

**Comment by KingsleyLiu-NV** [comment]
@wanchaol Thanks for the update. With #109306 and #109428, I do see some performance improvement for use case of convolution model (i.e., Shard [N, C, H, W] activations on W dimension, Replicate model weights):

| #GPUs | Local Input Shape  | Training Overhead | Forward Overhead | Backward Overhead | Tensor Type   |
|-------|-------------------|----------------------|---------------|---------------|---------------|
| 1     | (7, 3, 512, 1024) |  748  | 236 | 487 | Native Tensor as baseline |
| 1     | (7, 3, 512, 1024) | 1180 | 253 | 524 | DTensor       |
| 2     | (7, 3, 512, 512)  | 893 | 165 | 306 | DTensor       |
| 4     | (7, 3, 512, 256)  |  820 | 140 | 248 | DTensor       |
| 8     | (7, 3, 512, 128)  |  842 | 163 | 245 | DTensor       |

The overhead of optimizer step is reduced, though still not very performant. I have two questions here:

- Considering `#GPUs=1 with DTensor`, there does NOT require gradient averaging kernels (partial->replicate) actually, but its optimizer step overhead is still much higher than the baseline with DTensor. This can be verified with the simple script I provided in this issue. It would be appreciated if this can be more performant.

- Considering `#GPUs=2,4,8 with DTensor`,  the optimizer step overhead should be attributed to both **1)** gradient averaging overhead, as well as the **2)** unknown perf issue which also exists for the `#GPUs=1 with DTensor` case. Can you elaborate how to achieve "One can eagerly do the allreduce by installing gradient hooks to DTensor parameter so that the partial gradient convert to replicate gradient immediately after the backward gradient is available" based on the simple script I provided in this issue?
